Use shared code to create ShareableBitmap from a NativeImage

kkinnunen-apple opened the pull request:
#### 282a211f922289c79ed3bb3de6d0abe9fdd36a81
<pre>
Use shared code to create ShareableBitmap from a NativeImage
<a href="https://bugs.webkit.org/show_bug.cgi?id=298185">https://bugs.webkit.org/show_bug.cgi?id=298185</a>
<a href="https://rdar.apple.com/159590494">rdar://159590494</a>

Reviewed by NOBODY (OOPS!).

Use shared code to create ShareableBitmap from a NativeImage. Many places
would miss the opportunity to not draw the image, rather use the pixels
directly. Creating via a draw is bad because it may lose the color space,
because not all source color spaces are drawable.

In future, the draw path will not be needed for CG.

Reduces the usage of DestinationColorSpace NativeImage::colorSpace(),
which is a source of bugs and an oxymoron. In future, this will be
removed.

* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmap::createFromNativeImage):
(WebCore::ShareableBitmap::createFromImageDraw):
* Source/WebCore/platform/graphics/ShareableBitmap.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::createBitmapsFromNativeImage):
(WebCore::createBitmapsFromSVGImage):
(WebCore::decodeImageWithSize):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::createFrameImageAtIndex):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
(WebKit::RemoteGraphicsContextProxy::recordResourceUse):
* Source/WebKit/WebProcess/WebCoreSupport/ShareableBitmapUtilities.cpp:
(WebKit::createShareableBitmap):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadAndDecodeImage):
* Source/WebKit/WebProcess/cocoa/TextTrackRepresentationCocoa.mm:
(WebKit::WebTextTrackRepresentationCocoa::update):
* Tools/TestWebKitAPI/Resources/cocoa/icon-svg-16.tiff:
* Tools/TestWebKitAPI/Resources/cocoa/icon-svg-256.tiff:
Rebaselined the test expectation. Before, the .ico to .tiff conversion
would transfer the bitmaps with draws. This transformed straigth-alpha
CGImages from .ico data to premultiplied alpha, and then .tiff
was encoded with premultiplied alpha. Instead, encode the .tiffs
directly from the .ico CGImage data.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/282a211f922289c79ed3bb3de6d0abe9fdd36a81

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/170359 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/44282 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/37699 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/179733 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/128071 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/fabff4ff-5376-4b45-8b35-e9929dd09bb9) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/172225 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/45526 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/43914 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/132832 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/93636 "2 flakes 14 failures") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/41ee9087-1ff4-4704-9bc4-bfac62283b5e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/173318 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/36009 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/153261 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/113332 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/57a7d688-e12a-4832-947d-6c84d89a9252) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/34633 "Found 2 new test failures: imported/w3c/web-platform-tests/html/canvas/element/manual/wide-gamut-canvas/canvas-display-p3-drawImage.https.html imported/w3c/web-platform-tests/html/canvas/element/manual/wide-gamut-canvas/canvas-display-p3-pattern-image.html (failure)") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/32970 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/28417 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/145093 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/32659 "Found 2 new test failures: imported/w3c/web-platform-tests/html/canvas/element/manual/wide-gamut-canvas/canvas-display-p3-drawImage.https.html imported/w3c/web-platform-tests/html/canvas/element/manual/wide-gamut-canvas/canvas-display-p3-pattern-image.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/182318 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/35108 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/37136 "Found 2 new test failures: imported/w3c/web-platform-tests/html/canvas/element/manual/wide-gamut-canvas/canvas-display-p3-drawImage.https.html imported/w3c/web-platform-tests/html/canvas/element/manual/wide-gamut-canvas/canvas-display-p3-pattern-image.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/141280 "Passed tests") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/43939 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/141100 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/44628 "Built successfully") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/29644 "Found 2 new test failures: imported/w3c/web-platform-tests/html/canvas/element/manual/wide-gamut-canvas/canvas-display-p3-drawImage.https.html imported/w3c/web-platform-tests/html/canvas/element/manual/wide-gamut-canvas/canvas-display-p3-pattern-image.html (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/109071 "Built successfully") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/36368 "Passed tests") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/116510 "Built successfully") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/43138 "Built successfully") | [❌ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/7747 "") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/42544 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/42784 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/42673 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e045b62e831b890256a505954a8a5f42323b1083. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50138

kkinnunen-apple committed:
Use shared code to create ShareableBitmap from a NativeImage
https://bugs.webkit.org/show_bug.cgi?id=298185
rdar://159590494

Reviewed by NOBODY (OOPS!).

Use shared code to create ShareableBitmap from a NativeImage. Many places
would miss the opportunity to not draw the image, rather use the pixels
directly. Creating via a draw is bad because it may lose the color space,
because not all source color spaces are drawable.

In future, the draw path will not be needed for CG.

Reduces the usage of DestinationColorSpace NativeImage::colorSpace(),
which is a source of bugs and an oxymoron. In future, this will be
removed.

* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmap::createFromNativeImage):
(WebCore::ShareableBitmap::createFromImageDraw):
* Source/WebCore/platform/graphics/ShareableBitmap.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::createBitmapsFromNativeImage):
(WebCore::createBitmapsFromSVGImage):
(WebCore::decodeImageWithSize):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::createFrameImageAtIndex):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
(WebKit::RemoteGraphicsContextProxy::recordResourceUse):
* Source/WebKit/WebProcess/WebCoreSupport/ShareableBitmapUtilities.cpp:
(WebKit::createShareableBitmap):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadAndDecodeImage):
* Source/WebKit/WebProcess/cocoa/TextTrackRepresentationCocoa.mm:
(WebKit::WebTextTrackRepresentationCocoa::update):
* Tools/TestWebKitAPI/Resources/cocoa/icon-svg-16.tiff:
* Tools/TestWebKitAPI/Resources/cocoa/icon-svg-256.tiff:
Rebaselined the test expectation. Before, the .ico to .tiff conversion
would transfer the bitmaps with draws. This transformed straigth-alpha
CGImages from .ico data to premultiplied alpha, and then .tiff
was encoded with premultiplied alpha. Instead, encode the .tiffs
directly from the .ico CGImage data.

kkinnunen-apple committed:
Use shared code to create ShareableBitmap from a NativeImage
https://bugs.webkit.org/show_bug.cgi?id=298185
rdar://159590494

Reviewed by NOBODY (OOPS!).

Use shared code to create ShareableBitmap from a NativeImage. Many places
would miss the opportunity to not draw the image, rather use the pixels
directly. Creating via a draw is bad because it may lose the color space,
because not all source color spaces are drawable.

In future, the draw path will not be needed for CG.

Reduces the usage of DestinationColorSpace NativeImage::colorSpace(),
which is a source of bugs and an oxymoron. In future, this will be
removed.

* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmap::createFromNativeImage):
(WebCore::ShareableBitmap::createFromImageDraw): Deleted.
* Source/WebCore/platform/graphics/ShareableBitmap.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::createBitmapsFromNativeImage):
(WebCore::createBitmapsFromSVGImage):
(WebCore::decodeImageWithSize):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::createFrameImageAtIndex):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerManagerProxy.cpp:
(WebKit::RemoteMediaPlayerManagerProxy::bitmapImageForCurrentTime):
* Source/WebKit/WebProcess/GPU/graphics/RemoteNativeImageBackendProxy.cpp:
(WebKit::RemoteNativeImageBackendProxy::create):
* Source/WebKit/WebProcess/WebCoreSupport/ShareableBitmapUtilities.cpp:
(WebKit::createShareableBitmap):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadAndDecodeImage):
* Source/WebKit/WebProcess/cocoa/TextTrackRepresentationCocoa.mm:
(WebKit::WebTextTrackRepresentationCocoa::update):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/61b123cdbc6ad22eca7f140a62bf77f3211b529e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50138
